src/Core/Framework/DataAbstractionLayer/Cache/CachedEntityReader.php

```diff
@@ -55,6 +63,37 @@ public function read(EntityDefinition $definition, Criteria $criteria, Context $
         return $this->loadResultByIds($definition, $criteria, $context);
     }
 
+    /**
+     * See EntityHydrator for binary representation
+     * @param array|string $id
+     */
+    private static function makeIdToUniqueKey(EntityDefinition $definition, $id): string
+    {
+        if (is_array($id)) {
+            $primaryKeyFields = $definition->getPrimaryKeys();
+            $primaryKey = [];
+
+            /** @var Field $field */
+            foreach ($primaryKeyFields as $field) {
+                if ($field instanceof VersionField || $field instanceof ReferenceVersionField || !$field instanceof StorageAware) {
+                    continue;
+                }
+
+                $value = $id[$field->getStorageName()];
+
+                if ($field instanceof IdField || $field instanceof FkField) {
+                    $value = Uuid::fromHexToBytes($value);
+                }
+
+                $primaryKey[$field->getPropertyName()] = $field->getSerializer()->decode($field, $value);
```

Comment:
why decoding the value? values which provided to the entity reader should always be decoded previously